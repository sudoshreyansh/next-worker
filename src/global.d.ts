declare module '*.runtime.js' {
  const WorkerFactories: { [key: string]: () => Worker };
  export default WorkerFactories;
}