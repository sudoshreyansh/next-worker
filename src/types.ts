export enum WorkerStatus {
  LOADING,
  READY,
  ERROR,
  TERMINATED
}

export type WorkerState = {
  status: WorkerStatus;
  worker: Worker | undefined;
}

export type MessageListener = (e: MessageEvent) => void;