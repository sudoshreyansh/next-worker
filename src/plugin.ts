type NextJsWebpackConfig = (config: any, context: any) => any;

export default function NextWorkerPlugin(callback: NextJsWebpackConfig | undefined) {
  return (config, context) => {
    config.module.rules.push({
      test: /\.worker\.js$/,
      use: [{ loader: 'next-worker/build/loader' }]
    });

    if ( callback ) {
      config = callback(config, context);
    }

    return config;
  }
}