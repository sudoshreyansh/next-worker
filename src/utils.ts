import path from 'path';
import qs from 'querystring'

const NextWorkerDisableParamName = 'nextWorkerDisable';

export function escapeBackslash(val: string): string {
  return val.replace(/\\/g, '\\\\');
}

export function addLoaderDisableParams(file: string): string {
  const dirname = path.dirname(file);
  const basename = path.basename(file);
  const [filename, querystring] = basename.split('?');
  let params = {};

  if ( querystring ) {
    const existingParams = qs.parse(querystring);
    params = { ...existingParams };
  }

  params[NextWorkerDisableParamName] = 'true';

  const newQuerystring = qs.stringify(params);
  return path.join(dirname, filename + '?' + newQuerystring);
}

export function checkForLoaderDisable(file: string): boolean {
  const basename = path.basename(file);
  const querystring = basename.split('?')[1];
  const params = qs.parse(querystring);

  return params[NextWorkerDisableParamName] === 'true';
}

export function getWorkerChunkFile(file: string) {
  const dirname = path.dirname(file);
  const basename = path.basename(file);

  const matches = basename.match(/([a-z0-9-\._]+)\.worker(?:\.([a-z0-9-\._]+))?$/i);
  if ( !matches || matches.length === 0 )
    throw "Invalid worker file name. The file name should be of the form \"*.worker.[js|ts]\"";

  const basenamePrefix = matches[1];
  const basenameSuffix = matches[2] ? '.' + matches[2] : '';

  return path.join(dirname, basenamePrefix + '.worker.chunk' + basenameSuffix);
}

export function getWorkerSourceFile(chunk: string) {
  const dirname = path.dirname(chunk);
  const basename = path.basename(chunk);

  const matches = basename.match(/([a-z0-9-\._]+)\.worker\.chunk(?:\.([a-z0-9-\._]+))?$/i);
  if ( !matches || matches.length === 0 )
    throw "Invalid worker chunk. The file name should be of the form \"*.worker.chunk.[js|ts]\"";
  
  const basenamePrefix = matches[1];
  const basenameSuffix = matches[2] ? '.' + matches[2] : '';

  return path.join(dirname, basenamePrefix + '.worker' + basenameSuffix);
}