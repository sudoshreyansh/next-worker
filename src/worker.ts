import { WorkerFactory } from "./types";
import { useWorkerFactory } from "./hooks/useWorker";

export function generateWorkerHooks(workerFactory: WorkerFactory) {
  return {
    useWorker: useWorkerFactory(workerFactory)
  }
}
