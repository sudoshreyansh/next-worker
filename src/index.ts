/// <reference path="./types/global.d.ts" />

import plugin from './plugin';

export const NextWorkerPlugin = plugin;