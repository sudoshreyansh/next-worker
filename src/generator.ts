import path from 'path';
import { escapeBackslash } from "./utils";

export type WorkerEntry = {
  name: string;
  path: string;
};

export function generateRuntime(workerEntries: WorkerEntry[]): string {
  const escapedRelativeWorkerEntries = 
    workerEntries
      .map(e => ({
        name: e.name,
        path: 
          path.relative(__dirname, e.path)
            .split(path.sep)
            .join(path.posix.sep)
      }));
  
  const factories =
    escapedRelativeWorkerEntries
      .map(e => `'${e.name}': () => new Worker(new URL('${e.path}', import.meta.url))`)
      .join(', ');
  
  return (
    `export default { ${ factories } };`
  ) 
}