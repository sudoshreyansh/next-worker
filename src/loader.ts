import { generateWorkerRuntime } from './generator';
import { checkForLoaderDisable } from './utils';

export default function loader() {};

export function pitch(workerFilePath: string) {
  const disableStatus = checkForLoaderDisable(workerFilePath);
  if ( disableStatus ) return;

  return generateWorkerRuntime(workerFilePath);
}

