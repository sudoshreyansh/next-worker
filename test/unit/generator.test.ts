import { generateRuntime } from '../../src/generator';
import path from 'path';

describe('generateRuntime', () => {
  it('Snapshot testing of generated runtime', () => {
    const entries = [
      {
        name: 'example1',
        path: path.join(__dirname, 'example1.worker.js')
      },
      {
        name: 'example2',
        path: path.join(__dirname, 'hello world', 'example2.worker.js')
      },
    ]

    expect(generateRuntime(entries)).toMatchSnapshot();
  });
})