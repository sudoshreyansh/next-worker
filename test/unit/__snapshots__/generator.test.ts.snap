// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRuntime Snapshot testing of generated runtime 1`] = `"export default { 'example1': () => new Worker(new URL('../test/unit/example1.worker.js', import.meta.url)), 'example2': () => new Worker(new URL('../test/unit/hello world/example2.worker.js', import.meta.url)) };"`;
